src/librustc/util/ppaux.rs

```diff
@@ -907,6 +907,27 @@ impl<'tcx> fmt::Display for ty::TypeVariants<'tcx> {
             }
             TyTrait(ref data) => write!(f, "{}", data),
             ty::TyProjection(ref data) => write!(f, "{}", data),
+            ty::TyAnon(def_id, substs) => {
+                ty::tls::with(|tcx| {
+                    // Grab the "TraitA + TraitB" from `impl TraitA + TraitB`,
```

Comment:
The `+Trait` is eagerly parsed, so `F: Fn() -> (impl Trait + Send)`. However, `impl Trait` in a *bound* seems unlikely to me, and definitely not supported at this moment.
The precedence is this way for convenience - unlike trait objects, there's a prefix (`impl`), which solves all ambiguity *save* for what you just mentioned.
I suppose `-> impl Fn() -> impl Trait + Send` parses, as `-> impl Fn() -> (impl Trait + Send)`.